### DISCUSSION TITLE
While using pytest in -n mode, the window display is incomplete.

### DISCUSSION THREAD

**Discussion by ellieot** [seleniumbase/SeleniumBase#1860] [Q&A] [answered]
I run pytest like this:
`pytest -s .\test.py -n=2 --rs  --user-data-dir="home" --continue-on-collection-errors`

I think there should be two windows running at the same time, but I can only see one window appear.
I want to test if it works when 10 windows request data from the server together, so I need to run a set of code 10 times at the same time, and the way I can think of is to use -n, maybe I'm not using it the right way, I've checked the discussion and documentation, and I didn't find a clear usage of the -n command. I would like to ask for help.

and by the way, I want to log in at the first site and then use the same --rs command to synchronize the login information for all subsequent sites.

I know to use:
@pytest.mark.run(order=1)
This solves part of the problem, but if I want to run the same function 10 times, how do I ensure that only one window is logged in?

I asked chatGPT and it suggested that I can use:
@pytest.mark.parametrize("iteration", range(10))
Now my question is:
1. Is it normal to have only one window open while my command is -n=2?
2. How to use --rs command in this case, or is there a better way to login once in one function iteration?

**Comment by mdmintz** [comment]
Hi @ellieot, does your ``test.py`` file contain more than one test? Using ``-n=2`` will open 2 windows for test assuming that you have at least two tests in that file. (It's also possible that issues may arise when using multithreading with ``--user-data-dir=DIR``, as that can affect settings, and might not be thread-safe.) For tests that use ``--user-data-dir=Dir``, using ``--rs`` is the fastest way to speed up those tests safely.

Are you trying to run the same test multiple times? Parameterize them first. See the example: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_repeat_tests.py

Ordering with ``@pytest.mark.run(order=1)`` only guarantees the order for single-threaded tests. So if order is important, don't use ``-n=NUM`` with that. But you can still use ``--rs`` to reuse the session for all tests, and use ``@pytest.mark.run(order=1)`` to choose which tests run first.

**Reply by ellieot** [reply]
thanks for answering me, I tried it myself, and some problems were found
I have only one test function like:

```
class RecorderTest(BaseCase):
        def login(self):
            # ...login
            pass

        def test_recording(self):
            self.login()
            self.do_sth()
            ......
```
I need to run this one test method (test_recording()) 10 times simultaneously. and login for the first time.
So my idea is to rename the `login()` function to `test_login() ` function and add `@pytest.mark.run(order=1)` on it.
So my code will be like:
```
class RecorderTest(BaseCase):
        @pytest.mark.run(order=1)
        def test_login(self):
            # ...login
            pass

        @pytest.mark.parametrize("", [[]] * 2)
        def test_recording(self):
            self.do_sth()
            ......
```

and then using:
`pytest -s .\main.py -n=5 --rs`
but still just one window appeard, I asked the chatGPT, and it cant answer the question well...

**Reply by mdmintz** [reply, marked as answer]
Since your test class inherits ``BaseCase``, you must use ``@parameterized.expand([[]] * 2)`` instead of ``@pytest.mark.parametrize("", [[]] * 2)``

Example:

```python
from parameterized import parameterized
from seleniumbase import BaseCase


class RepeatTests(BaseCase):
    @parameterized.expand([[]] * 2)
    def test_repeat_this_test_with_parameterized(self):
        self.open("seleniumbase.github.io")
        self.click('a[href="help_docs/method_summary/"]')
        self.assert_text("API Reference", "h1")
```
